File: backend/app/schemas/purchase_benchmarks.py
"""Schemas — Sprint 8 purchase benchmarks."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class BenchmarkRunCreate(BaseModel):
    purchase_invoice_id: UUID
    comparison_mode: str | None = None


class BenchmarkReprocessRequest(BaseModel):
    reason: str = Field(min_length=3, max_length=2000)


class BenchmarkReferenceOverrideRequest(BaseModel):
    reference_datetime: datetime
    reason: str = Field(min_length=3, max_length=2000)


class BenchmarkDistributorOverrideRequest(BaseModel):
    distributor_id: UUID
    reason: str = Field(min_length=3, max_length=2000)


class BenchmarkParametersUpsert(BaseModel):
    absolute_tolerance_per_liter: str = "0.005"
    percentage_tolerance: str = "0.001"
    allow_low_confidence_reference: bool = True
    default_comparison_mode: str = "DELIVERED_COST"


class BenchmarkRunListItem(BaseModel):
    id: UUID
    purchase_invoice_id: UUID
    station_id: UUID
    status: str
    comparison_mode: str
    reference_datetime: datetime | None = None
    reference_source: str
    reference_confidence: str
    item_count: int
    benchmarked_item_count: int
    opportunity_amount: str | None = None
    cost_variance_amount: str | None = None
    created_at: datetime
    finished_at: datetime | None = None
    snapshot_hash: str | None = None


class BenchmarkRunListResponse(BaseModel):
    items: list[BenchmarkRunListItem]
    total: int
    page: int
    page_size: int


class BenchmarkItemResponse(BaseModel):
    id: UUID
    group_key: str
    canonical_product_id: UUID | None = None
    actual_distributor_id: UUID | None = None
    volume_liters: str
    actual_delivered_cost: str
    actual_delivered_cost_per_liter: str | None = None
    benchmark_status: str
    decision_result: str
    best_quote_id: UUID | None = None
    best_distributor_id: UUID | None = None
    benchmark_cost_per_liter: str | None = None
    cost_variance_per_liter: str | None = None
    cost_variance_amount: str | None = None
    opportunity_amount: str | None = None
    actual_advantage_amount: str | None = None
    actual_distributor_rank: int | None = None
    candidate_count: int
    eligible_candidate_count: int
    exclusion_reasons: list | None = None
    warnings: list | None = None


class BenchmarkCandidateResponse(BaseModel):
    id: UUID
    quote_id: UUID
    quote_item_id: UUID
    distributor_id: UUID
    eligibility_status: str
    ranking_position: int | None = None
    is_best: bool
    delivered_cost_per_liter: str | None = None
    financial_equivalent_per_liter: str | None = None
    blocking_reasons: list | dict | None = None


class BenchmarkRunDetailResponse(BaseModel):
    id: UUID
    purchase_invoice_id: UUID
    station_id: UUID
    status: str
    comparison_mode: str
    reference_datetime: datetime | None = None
    reference_source: str
    reference_confidence: str
    trigger_type: str
    reprocess_of_run_id: UUID | None = None
    reprocess_reason: str | None = None
    item_count: int
    benchmarked_item_count: int
    warning_count: int
    error_count: int
    actual_total_cost: str
    benchmark_total_cost: str | None = None
    cost_variance_amount: str | None = None
    opportunity_amount: str | None = None
    actual_advantage_amount: str | None = None
    snapshot_hash: str | None = None
    input_snapshot: dict | None = None
    output_snapshot: dict | None = None
    started_at: datetime
    finished_at: datetime | None = None
    created_at: datetime
    items: list[BenchmarkItemResponse] = []


class BenchmarkSummaryResponse(BaseModel):
    purchase_group_count: int
    benchmarked_group_count: int
    purchased_volume_liters: str
    benchmarked_volume_liters: str
    coverage_volume_ratio: str | None = None
    actual_total_cost: str
    benchmark_total_cost: str | None = None
    cost_variance_amount: str | None = None
    opportunity_amount: str | None = None
    best_or_tied_count: int


class BenchmarkCoverageResponse(BaseModel):
    total_groups: int
    total_volume_liters: str
    total_value: str
    by_status: dict


class BenchmarkDataQualityResponse(BaseModel):
    unmapped_product_count: int
    unmapped_supplier_warning_count: int
    missing_cost_count: int
    missing_volume_count: int
    reference_unavailable_count: int
    no_quotes_count: int
    no_eligible_count: int
    not_comparable_count: int
    low_confidence_count: int


class BenchmarkOpportunityRow(BaseModel):
    benchmark_item_id: UUID
    purchase_invoice_id: UUID
    station_id: UUID
    canonical_product_id: UUID | None = None
    volume_liters: str
    opportunity_amount: str | None = None
    cost_variance_per_liter: str | None = None
    decision_result: str


class BenchmarkParametersResponse(BaseModel):
    id: UUID
    absolute_tolerance_per_liter: str
    percentage_tolerance: str
    allow_low_confidence_reference: bool
    default_comparison_mode: str
    valid_from: datetime
    valid_until: datetime | None = None
    active: bool

File: frontend/src/api/purchase-benchmarks.ts
import { apiFetch } from "./client";

export type BenchmarkSummary = {
  purchase_group_count: number;
  benchmarked_group_count: number;
  purchased_volume_liters: string;
  benchmarked_volume_liters: string;
  coverage_volume_ratio: string | null;
  actual_total_cost: string;
  benchmark_total_cost: string | null;
  cost_variance_amount: string | null;
  opportunity_amount: string | null;
  best_or_tied_count: number;
};

export type BenchmarkCoverage = {
  total_groups: number;
  total_volume_liters: string;
  total_value: string;
  by_status: Record<string, unknown>;
};

export type BenchmarkDataQuality = {
  unmapped_product_count: number;
  unmapped_supplier_warning_count: number;
  missing_cost_count: number;
  missing_volume_count: number;
  reference_unavailable_count: number;
  no_quotes_count: number;
  no_eligible_count: number;
  not_comparable_count: number;
  low_confidence_count: number;
};

export type BenchmarkOpportunity = {
  benchmark_item_id: string;
  purchase_invoice_id: string;
  station_id: string;
  canonical_product_id: string | null;
  volume_liters: string;
  opportunity_amount: string | null;
  cost_variance_per_liter: string | null;
  decision_result: string;
};

export type BenchmarkRunDetail = {
  id: string;
  purchase_invoice_id: string;
  station_id: string;
  status: string;
  comparison_mode: string;
  reference_datetime: string | null;
  reference_source: string;
  reference_confidence: string;
  item_count: number;
  benchmarked_item_count: number;
  actual_total_cost: string;
  benchmark_total_cost: string | null;
  cost_variance_amount: string | null;
  opportunity_amount: string | null;
  actual_advantage_amount: string | null;
  snapshot_hash: string | null;
  items: Array<{
    id: string;
    group_key: string;
    volume_liters: string;
    actual_delivered_cost_per_liter: string | null;
    benchmark_cost_per_liter: string | null;
    cost_variance_per_liter: string | null;
    decision_result: string;
    benchmark_status: string;
    opportunity_amount: string | null;
    actual_distributor_rank: number | null;
  }>;
};

function rangeQs(dateFrom?: string, dateTo?: string) {
  const sp = new URLSearchParams();
  if (dateFrom) sp.set("date_from", dateFrom);
  if (dateTo) sp.set("date_to", dateTo);
  const s = sp.toString();
  return s ? `?${s}` : "";
}

export function fetchBenchmarkSummary(dateFrom?: string, dateTo?: string) {
  return apiFetch<BenchmarkSummary>(`/api/v1/analytics/purchase-benchmarks/summary${rangeQs(dateFrom, dateTo)}`);
}

export function fetchBenchmarkCoverage(dateFrom?: string, dateTo?: string) {
  return apiFetch<BenchmarkCoverage>(`/api/v1/analytics/purchase-benchmarks/coverage${rangeQs(dateFrom, dateTo)}`);
}

export function fetchBenchmarkDataQuality(dateFrom?: string, dateTo?: string) {
  return apiFetch<BenchmarkDataQuality>(
    `/api/v1/analytics/purchase-benchmarks/data-quality${rangeQs(dateFrom, dateTo)}`,
  );
}

export function fetchBenchmarkOpportunities(dateFrom?: string, dateTo?: string) {
  return apiFetch<BenchmarkOpportunity[]>(
    `/api/v1/analytics/purchase-benchmarks/opportunities${rangeQs(dateFrom, dateTo)}`,
  );
}

export function fetchBenchmarkRun(id: string) {
  return apiFetch<BenchmarkRunDetail>(`/api/v1/purchase-benchmarks/runs/${id}`);
}

export function runInvoiceBenchmark(invoiceId: string) {
  return apiFetch<BenchmarkRunDetail>(`/api/v1/fuel-purchase-invoices/${invoiceId}/benchmark`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ purchase_invoice_id: invoiceId }),
  });
}

export function fetchLatestInvoiceBenchmark(invoiceId: string) {
  return apiFetch<BenchmarkRunDetail>(`/api/v1/fuel-purchase-invoices/${invoiceId}/latest-benchmark`);
}

export function reprocessBenchmarkRun(runId: string, reason: string) {
  return apiFetch<BenchmarkRunDetail>(`/api/v1/purchase-benchmarks/runs/${runId}/reprocess`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ reason }),
  });
}
